W-17700088-outbound-policies-fa (#757)
mulesoft/docs-gateway: gateway/1.9/modules/ROOT/pages/policies-outbound-circuit-breaker.adoc

= Circuit Breaker Outbound Policy

[%autowidth.spread,cols="a,a"]
|===
>s| Policy name | Rate Limiting
>s| Summary      | Prevents an API from making outbound requests to an unresponsive or failing service.
>s| Category | Quality of Service
>s| First Flex Gateway version available | v1.2.0
.4+>.^s| Returned Status Codes
|502 - Bad Gateway, requests are blocked due to circuit breaker activation
|===

== Summary

The Circuit Breaker policy helps prevent an API from making outbound requests to an unresponsive or failing service. By monitoring response failures, this policy automatically blocks requests to the failing service for a defined period, preventing unnecessary load on both the API and the backend service.

When you configure the Rate Limiting policy, you can specify any number of pairs of quota (number of requests) and time window (time period) values.

[NOTE]
====
A Rate Limiting policy applied to a Flex Gateway API is scoped to replicas, not the gateway.
====

== Configuring Policy Parameters

=== Flex Gateway Local Mode

> In Local Mode, you apply the Rate Limiting policy to your API via declarative configuration files. Refer to the following policy definition and table of parameters:

Reviewer: rate limit?

----
- policyRef:
    name: circuit-breaker-flex
  config:

>       - maximumConnections: <int> // OPTIONAL

Reviewer: thresholds:
   max ......
  
```
apiVersion: gateway.mulesoft.com/v1alpha1
kind: PolicyBinding
metadata:
  name: upstream-policy
spec:
  targetRef:
    Kind: Service
    name: upstream
  policyRef:
    name: circuit-breaker-flex
  config:
    thresholds:
      maxConnections: 1
      maxPendingRequests: 1
      maxRequests: 50
      maxRetries: 2
      maxConnectionPools: 50
```

        maximumPendingRequests: <int> // OPTIONAL
        maximumRequests: <int> // OPTIONAL
        maximumRetries: <int> // OPTIONAL, 
        maximumConnectionPools: <int> // OPTIONAL
----

[%header%autowidth.spread,cols="a,a,a"]
|===
|Parameter | Required or Optional | Description

| `maximumConnections`
| Optional
| Maximum number of connections to all hosts in an upstream.

| `maximumPendingRequests`
| Optional
| The maximum number of requests queued while waiting for a connection.

| `maximumRequests`
| Optional
| The maximum number of requests that can be outstanding to all hosts in a cluster.
| `maximumretries`
| Optional
| The maximum number of retries in a cluster.
| `maximumConnectionPools`
| Optional
| The maximum number of connection pools that can be concurrently initiated.

|===

==== Resource Configuration Example

> The following example defines a rate limit of three requests every six seconds:

Reviewer: rate limit?

----

> - policyRef:

Reviewer: is this for the configuration resource? in that case here you have an example
```
---
apiVersion: gateway.mulesoft.com/v1beta1
kind: Configuration
metadata:
  name: circuit-breaker
  namespace: test-ns
spec:
  circuitBreaker:
    thresholds:
      maxConnections: 1
      maxPendingRequests: 1
      maxRequests: 1
      maxRetries: 1
      maxConnectionPools: 1
```

= Circuit Breaker Policy

[%autowidth.spread,cols="a,a"]
|===
>s| Policy name | Circuit Breaker

> >s| Summary | Prevents an upstream from receiving too many requests or connection

Reviewer 2: ```suggestion
>s| Summary | Prevents an upstream from receiving too many requests or connections
```

>s| Category | Quality of Service
>s| First Flex Gateway version available | v1.9.0
.4+>.^s| Returned Status Codes
|503 - Service Unavailable
|===

== Summary

> The Circuit Breaker policy helps prevent an upstream service from receiving too many requests or connections at a time. If any defined maximums are reached, Flex Gateway rejects the request or connection with a `503 - Service Unavailable` error code.

Reviewer 2: ```suggestion
The Circuit Breaker policy prevents upstream services from receiving too many requests or connections at a time. If the defined maximums are reached, Flex Gateway returns a `503 - Service Unavailable` error.
```

> For Managed Flex Gateways or Connected Mode, if different API instances request the same upstream, the limits are independent, so each API instance can use a full maximum. For Local Mode, the limits are not independent, so different API instances share the maximums for an upstream. For all modes, maximums are not distributed across replicas, so each Gateway can use can use a full maximum.

Reviewer 2: ```suggestion
For Managed Flex Gateways or Connected Mode, each API instance has its own independent limit when requesting the same upstream, allowing each to use the full maximum. In Local Mode, the limits are shared among different API instances, so they must all adhere to the same upstream maximum. For all modes, maximums are not distributed across replicas; each Gateway can use the full maximum.
```

> The Circuit Breaker policy is an extension of the Envoy upstream clusters circuit breaking feature. To learn more about the Envoy's circuit breaking, see https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking.html[Circuit Breaking].

Reviewer 2: ```suggestion
The Circuit Breaker policy is an extension of the Envoy upstream clusters circuit breaking feature. To learn more about the Envoy's circuit breaking, see https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking.html[Circuit Breaking] in the Envoy documentation.
```

kind: PolicyBinding  
metadata:  
  name: upstream-policy  
spec:  
  targetRef:  
    Kind: Service  
    name: upstream  
  policyRef:  
    name: circuit-breaker-flex  
  config:  
    thresholds:  
      maxConnections: <number> // OPTIONAL 
      maxPendingRequests: <number> // OPTIONAL  
      maxRequests: <number> // OPTIONAL
      maxRetries: <number> // OPTIONAL  
      maxConnectionPools: <number> // OPTIONAL
----

NOTE: For Local Mode, the limits are not independent, so different API instances share the maximums for an upstream.

[%header%autowidth.spread,cols="a,a,a"]
|===
|Parameter | Required or Optional | Description

| `maximumConnections`
| Optional
| Maximum number of connections per API to an upstream.

| `maximumPendingRequests`
| Optional
| Maximum number of queued requests while waiting for a connection.

| `maximumRequests`
| Optional
| Maximum number of outstanding requests to an upstream.

| `maximumRetries`
| Optional
| Maximum number of outstanding retries.

| `maximumConnectionPools`
| Optional
| Maximum number of connection pools concurrently initiated to an upstream.

|===

==== Resource Configuration Example

> The following example defines a  of three requests every six seconds:

Reviewer 2: ```suggestion
This example defines a policy of three requests every six seconds:
```

----
apiVersion: gateway.mulesoft.com/v1beta1  
kind: Configuration  
metadata:  
  name: circuit-breaker  
  namespace: test-ns  
spec:  
  circuitBreaker:  
    thresholds:  
      maxConnections: 1  
      maxPendingRequests: 1  
      maxRequests: 1  
      maxRetries: 1  
      maxConnectionPools: 1  
----

=== Flex Gateway Connected Mode

When you apply the Rate Limiting policy to your API from the UI, you can configure the following parameters:

[%header%autowidth.spread,cols="a,a,a"]
|===
|Parameter | Required or Optional | Description

| Maximum connections
| Optional
| Maximum number of connections per API to an upstream.

| Maximum pending requests
| Optional
| Maximum number of queued requests while waiting for a connection.

| Maximum requests
| Optional
| Maximum number of outstanding requests to an upstream.

| Maximum retries
| Optional
| Maximum number of outstanding retries.

| Maximum connection pools
| Optional
| Maximum number of connection pools concurrently initiated to an upstream.

|===

> NOTE: For Managed Flex Gateways or Connected Mode, if different API instances request the same upstream, the limits are independent, so each API instance can use a full maximum.

Reviewer 2: ```suggestion
NOTE: For Managed Flex Gateways or Connected Mode, each API instance has its own independent limit when requesting the same upstream, allowing each to use the full maximum.
```
